src/Storages/MergeTree/IMergeTreeDataPart.cpp

```diff
@@ -1285,6 +1302,85 @@ void IMergeTreeDataPart::loadRowsCount()
     }
 }
 
+void IMergeTreeDataPart::readExistingRowsCount() const
+{
+    if (!supportLightweightDeleteMutate() || !hasLightweightDelete() || !storage.getSettings()->exclude_deleted_rows_for_part_size_in_merge
+        || existing_rows_count < rows_count || !getMarksCount())
+        return;
+
+    std::lock_guard lock(existing_rows_count_mutex);
```

Comment:
do we need to protect this with mutex? I'm not quite sure
@KochetovNicolai could you please take a look?